README.md
[![Documentation Status](https://readthedocs.org/projects/customxepr/badge/?version=latest)](https://customxepr.readthedocs.io/en/latest/?badge=latest)

**Warning:** Version 2.3.3 is the last release that supports Python 2.7. All newer 
releases only support Python 3.6 and higher.

# CustomXepr

A Python instrument controller and GUI for Bruker E500 EPR spectrometers, MercuryiTC
temperature controllers and Keithley 2600 series source measurement units. CustomXepr
relies on the python drivers [keithley2600](https://github.com/OE-FET/keithley2600) and
[mercuryitc](https://github.com/OE-FET/mercuryitc) and the respective user interfaces
[keithleygui](https://github.com/OE-FET/keithleygui) and
[mercurygui](https://github.com/OE-FET/mercurygui) for functionality regarding the
Keithley 2600 and MercuryiTC instruments.

![Screenshot of CustomXepr GUI](/screenshots/CustomXepr_all.png)

## Overview

CustomXepr for Linux and macOS enables the interaction with all instruments involved in
field induced electron spin resonance (FI-EPR) measurements: the Bruker E500 spectrometer,
through Bruker's Xepr Python API, the Oxford Instruments MercuryiTC temperature
controller, and the Keithley 2600 series of source measurement units (SMUs). The program
suite is structured into drivers and user interfaces for individual instruments (external
packages), CustomXepr's main class which provides higher level functions that often
combine functionality from multiple instruments, and a manager which handles the
scheduling of experiments.

The aim of CustomXepr is twofold: First and foremost, it enables the user to automate and
schedule full measurement plans which may run for weeks without user input. Second, it
complements the functionality of Bruker's Xepr control software. This includes for
instance powerful logging capabilities, a more accurate determination of the cavity's
Q-value from its frequency response, more reliable tuning of the cavity, the ability to
re-tune the cavity during long-running measurements, logging of the cryostat temperature
during measurements, and many more. Low level functionality and communication with the
spectrometer remains with Xepr.

![CustomXepr structure](/screenshots/CustomXepr_structure.png)

Finally, CustomXepr fully supports Bruker BES3T data files (DSC, DTA, etc). The
`customxepr.experiment.XeprData` class enables loading, plotting, modifying and saving such
data files from regular, 2D or pulsed experiements. It also supports reading and plotting
the pulse sequences used to acquire the data, as saved in the DSC file. More information
is provided in the API documentation for the
[XeprData](https://customxepr.readthedocs.io/en/latest/api/xepr_dataset.html#experiment.xepr_dataset.XeprData)
class.

## Installation

Make sure that you have PyQt or PySide installed on your system (all other dependencies
will be installed automatically). Then install CustomXepr by running in a terminal:
```
$ pip install git+https://github.com/OE-FET/customxepr
```

## Instrument communication

CustomXepr communicates with with the Keithley and MercuryiTC through NI-VISA or pyvisa-py
and is therefore independent of the actual interface, e.g., Ethernet, USB, or GPIB.
Connections to the EPR spectrometer are handled through Bruker's Xepr Python API.

## Usage

CustomXepr can be run interactively from a Jupyter console, or as a standalone program. In
the latter case, it will create its own internal Jupyter console for the user to run
commands.

You can start CustomXepr from a Python command prompt as follows:
```python
>>> from customxepr import run
>>> customXepr, xepr, mercury, keithley, ui = run()
```
If executed from an Jupyter console, this will automatically start the integrated Qt event
loop and run in interactive mode. To start CustomXepr from the console / terminal, run
`customxepr`.

CustomXepr has a user interface which displays all jobs waiting in the queue, all results
returned from previous jobs, and all logging messages. Common tasks such as pausing,
aborting and clearing jobs, plotting and saving returned data, and setting temperature
stability tolerances can be performed through the interface itself. However, apart from
tuning the cavity and reading a Q-factor, all jobs  must be scheduled programmatically
through the provided Jupyter console.

## Job-scheduling

CustomXepr's core consists of functions for preconfigured tasks, such as changing the
cryostat temperature, recording a transfer curve, performing a preconfigured EPR
measurement. For instance, `customXepr.setTemperature(110)` tells the MercuryiTC to change
its temperature set-point to 110 K and waits until the latter is reached and maintained
with the desired stability (default: ±0.1 K for 120 sec). It also adjusts the helium flow
if necessary and warns the user if the target temperature cannot be reached within the
expected time. `customXepr.runExperiment(powersat)` will run the preconfigured EPR
measurement "powersat" while tuning the cavity between scans and monitoring the
temperature stability during the measurement.

Such built in jobs are not performed immediately but are queued and run after the
successful completion of the previous jobs. Any data returned by a job, such as a transfer
curve or a cavity mode picture, will be kept in a result queue and saved to a specified
file if requested. If the returned object has `save` and `plot` methods implemented, one
can right-click on its entry in the GUI to plot the data or save it to the hard drive.

CustomXepr functions that are expected to run for longer than 1 sec can gracefully abort
upon user request without leaving the setup in an inconsistent state.

![Screenshot of CustomXepr GUI](/screenshots/CustomXepr_jobs.png)

In addition, the queuing system can be used to manually schedule any user-specified jobs,
related or unrelated to the EPR setup and its ancillary equipment. This can be done with
the `queued_exec` decorator from `customxepr.manager`:

```python
>>> import time
>>> from customxepr.manager import Manager
>>> manager = Manager()
>>> # create test function
>>> @manager.queued_exec
... def test_func(*args):
...     # do something
...     for i in range(0, 10):
...         time.sleep(1)
...		    # check for requested aborts
...         if manager.abort.is_set():
...             break
...     return args  # return input arguments
>>> # run the function: this will return immediately
>>> test_func('test succeeded')
```

The result returned by `test_func` can be retrieved from the result queue as follows:

```python
>>> result = manager.result_queue.get()  # blocks until result is available
>>> print(result)
test succeeded
```

More information regarding the manual scheduling of experiments can be found
[here](https://customxepr.readthedocs.io/en/latest/api/manager.html#manager.Manager).

## Logging and error handling

The detection and escalation of possible problems is key to enabling unattended
measurements. Otherwise the user may come back after two days expecting a completed
measurement cycle, only to see that the helium dewar was emptied a day ago or that the
program got stuck asking the user if it should really override a data file. CustomXepr
therefore includes logging capabilities to track the progress of jobs and notify the user
of potential problems.

All CustomXepr methods release logging messages during their execution which may have the
levels "status", "info", "warning", and "error". Status notifications will only be shown
in the user interface and typically contain information about the progress of a job
(number of completed scans in an EPR measurement, countdown until the temperature is
stable, etc). Info notifications typically contain information about the beginning or
completion of a job (e.g., "Waiting for temperature to stabilize.", "All scans
complete."), and potentially useful information about how the job was completed (e.g.,
"Temperature stable at 120.01±0.02 K during scans.").

Warning notifications are logged when CustomXepr believes that there may be a problem
which requires user intervention, for instance if a job is taking significantly longer
than expected, or if the gas flow required to maintain a certain temperature is unusually
high. Finally, error messages are released if CustomXepr is unable to proceed with a job,
in which case it will abort and pause all pending jobs. Such errors may include loss of
communication with an instrument, repeated strong temperature fluctuations during an EPR
measurement, etc.

By default, all messages of level "info" and higher are saved to a log file in the user's
home directory and messages of level "warning" and higher are sent as an email to the
user's address. In addition, temperature readings are saved to a log file every 5 min,
allowing the user to retrospectively confirm the temperature stability during
measurements.

![Screenshot of CustomXepr GUI](/screenshots/CustomXepr_log.png)

## Mercury controls

CustomXepr includes a higher-level worker thread which regularly queries the MercuryiTC
for its sensor readings and provides a live stream of this data to other parts of the
software. This prevents individual functions from querying the MercuryiTC directly and
causing unnecessary overhead.

The user interface for the cryostat plots historic temperature readings going back up to
24 h and provides access to relevant temperature control settings such as gas flow, heater
power, and ramping speed while lower-level configurations such as calibration tables must
be changed programmatically through the provided Jupyter console.

The MercuryiTC user interface and driver have been split off as separate packages
[mercuryitc](https://github.com/OE-FET/mercuryitc) and [mercurygui](https://github.com/OE-
FET/mercurygui).

![Screenshot of Mercury GUI](/screenshots/MercuryGUI.png)

## Keithley controls

As with the cryostat, CustomXepr includes a high-level user interface for Keithley 2600
series instruments which allows the user to configure, record and save voltage sweeps such
as transfer and output measurements. Since there typically is no need to provide a live
stream of readings from the Keithley, the data from an IV-curve is buffered locally on the
instrument and only transferred to CustomXepr after completion of a measurement.

The Keithley 2600 user interface and driver have been split off as separate packages
[keithley2600](https://github.com/OE-FET/keithley2600) and
[keithleygui](https://github.com/OE-FET/keithleygui).

![Screenshot of Keithley GUI](/screenshots/KeithleyGUI.png)

## Example code

A measurement script which cycles through different temperatures and records EPR spectra
and transfer curves at each step reads as follows:

```python
# start customxepr
from customxepr.startup import run

customXepr, xepr, mercury, keithley, _ = run(gui=False)

# get preconfigured experiment from Xepr
exp = xepr.XeprExperiment('Experiment')

# set up different modulation amplitudes in Gauss for different temperatures
modAmp = {5: 3, 50: 2, 100: 1, 150: 1, 200: 1, 250: 1.5, 300: 2}

# specify folder to save data
folder = '/path/to/folder'
title = 'my_sample'

for T in [5, 50, 100, 150, 200, 250, 300]:
	# =================================================================
	# Prepare temperature
	# =================================================================
	customXepr.setTemperature(T)        # set desired temperature
	customXepr.customtune()             # tune the cavity
	customXepr.getQValueCalc(folder, T) # measure and save the Q factor

	# =================================================================
	# Perform FET measurements
	# =================================================================
	# generate file name for transfer curve
	transfer_file = '{}/{}_{}K_transfer.txt'.format(folder, title, T)
	# record default transfer curve and save to file
	customXepr.transferMeasurement(path=transfer_file)

	# =================================================================
	# Perform EPR measurements at Vg = -70V and Vg = 0V
	# =================================================================
	for Vg in [0, -70]:
        customXepr.setVoltage(Vg, smu='smua')  # bias gate
        # perform preconfigured EPR measurement, save to 'esr_path'
        esr_file = '{}/{}_{}K_Vg_{}V.txt'.format(folder, title, T, Vg)
        customXepr.runXeprExperiment(exp, path=esr_file, ModAmp=modAmp[T])
        customXepr.setVoltage(0, smu='smua')  # set gate voltage to zero

customXepr.setStandby()  # ramp down field and set MW bridge to standby
```

In this code, all functions belonging to CustomXepr will be added to the job queue and
will be carried out successively such that, for instance, EPR measurements will not start
while the temperature is still being ramped.

## Email notifications

By default, email notifications are sent from 'customxepr@outlook.com'. CustomXepr at the moment
provides no way to modify the email settings via the user interface, but you can set them manually
in the config file in your home directory: '~/.CustomXepr/CustomXepr.ini'. Changes will be applied
when restarting CustomXper.

By default, the relevant section in the config file reads:

```ini
[SMTP]
mailhost = localhost
port = 0
fromaddr = ss2151@cam.ac.uk
credentials =
secure =
```

Authentication credentials can be specified as a tuple `(username, password)`. To specify the use
of a secure protocol (TLS), pass in a tuple for the `secure` argument. This will only be used when
authentication credentials are supplied. The tuple will be either an empty tuple, or a single-value
tuple with the name of a keyfile, or a 2-value tuple with the names of the keyfile and certificate
file.

## Requirements

*System requirements:*

- Linux or macOS
- Python 2.7 or 3.6 and higher

*Recommended:*

- Bruker Xepr software with Python XeprAPI (required for EPR related functions),
  a Python 3 version of the API is available [here](https://github.com/OE-FET/XeprAPI)
- Postfix - mail transfer agent (required for email notifications from localhost)

*Required python modules:*
- PyQt5 >= 5.9

*Recommended python modules:*
- pyusb (only when using pyvisa-py backend)
- pyserial (only when using pyvisa-py backend)

## Acknowledgements

- Config modules are based on the implementation from [Spyder](https://github.com/spyder-ide).
- Scientific spin boxes are taken from [qudi](https://github.com/Ulm-IQO/qudi).


CHANGELOG.md
## v3.1.3

#### Fixed:

- Fixed spurious notifications from log messages emitted outside of customxepr. 

## v3.1.2

#### Added:

- Added support to parse and plot the pulse sequence information stored in the Bruker DSC
  file in the tables Psd1 to Psd36:

  ```Python
  >>> from customxepr.experiment import XeprData
  >>> dset = XeprData("path")
  >>> dset.pulse_sequence.plot()
  >>> print(dset.pulse_sequence.pulse_channels[5])
  <PulseChannel(Psd6: +x)>
  >>> print(dset.pulse_sequence.pulse_channels[5].pulses)
  [<Pulse(position=276, length=4)>, <Pulse(position=676, length=44)>]
  ```
- Added a keyword argument `settling_time` to `runXeprExperiment` to wait between
  individual scans. This can be useful in case of temperature fluctuations between
  scans.

#### Changed:

- Remove custom `exit_customxepr()` function. Instead, close the console to exit
  CustomXepr or run `exit` when started from an IPython or Jupyter shell.
- Increase default maximum cooling temperature to 20°C.
- Remember the location of the console window between restarts.
- Set current directory of Xepr to directory of last-saved data.
- The Qt event loop is no longer started on every import but just when actually starting
  the GUI or plotting a figure.
- `getValueXepr` no longer accepts a path argument.
- `getQValueCalc` no longer saves two files but the mode picture only. The given path
  must now be the full file path. The temperature data will now be saved together with
  the mode picture.
- `getQValueCalc` no longer accepts a temperature argument. Instead, the temperature
  will be read from the Mercury if available.
- The `ModePicture` class now accepts arbitrary metadata in dictionary form. This data
  will be saved as tab-delimited metadata in the mode picture file.

#### Fixed:

- Fixed an issue when saving a dateset through Xepr when the path contains whitespace or
  backslashes. The path is now quoted with `shlex.quote` before being passed to Xepr.

## v3.1.1

#### Changed:

- Only show single-line previews of results in GUI.
- Improve warnings for too high gasflow.
- Allow overriding default tolerance and wait time in `waitTemperatureStable`.

#### Added:

- Monitor cooling water temperature during measurements and interrupt ESR measurements if
  too high.
- Added config file entries for the maximum cooling temperature and the temperature sensor
  names for the ESR and cooling water.

#### Fixed:

- Fixes an issue in `customxepr.experiment.xepr_dataset.XeprParam` where the `unit`
  attribute would return the value instead of the unit.

## v3.1.0

#### Added:

- Added API similar to `concurrent.futures.Future` to `customxepr.manager.Experiment`.
- Return an experiment instance from all async customxepr method calls. This instance can
  be used to get the result directly, instead of getting it from the result queue. Usage:

  ```Python
  future = customXepr.runXeprExperiment(...)
  result = future.result()  # will block until result is available
  ```

#### Changed:

- `CustomXepr.tune` and `CustomXepr.finetune` can now be aborted.
- `ModePicture.load(...)` no longer returns the raw data but rather populates the class
  attributes of `ModePicture` with the loaded data.
- `XeprData` now shows the experiment title in its string representation.
- Changed the default email address for notifications to physics-oe-esr-owner@lists.cam.ac.uk.
- Improvements to documentation.
- Require PyQt5 instead of qtpy / PySide2.

#### Fixed:

- Fixes an issue where `CustomXepr.tune` and `CustomXepr.finetune` would return
  immediately, even if tuning was not complete.

## v3.0.0

This release drops support for Python 2.7. Only Python 3.6 and higher are supported.

#### Added:

- Method `CustomXepr.getExpDuration` to estimate the duration of an Xepr experiment.
- Added synchonous functions for all of CustomXepr's asynchronous functions (which will be
  queued). These are automatically generated and end with the suffix "_sync".
- Added `shape` attribute to `XeprData` class.

#### Changed:

- Changed the abort behaviour of a measurement: Instead of finishing the current scan and
  pausing afterwards, the scan is aborted immediately.
- Renamed `setDrainCurrent` to `setCurrent` and `setGateVoltage` to `setVoltage`.
- `setVoltage` no longer turns the other SMUs off.
- Optimized the truncation of long items in the list of running expriements.
- Changed the priority of locations to search for the XeprAPI:
    1) path from the environment variable `XEPR_API_PATH` (if set)
    2) installed python packages
    3) pre-installed version from Xepr
- Renamed `applyCurrent` to `setCurrent`.
- The `queued_exec` decorator is now an attribute of `customXepr.manager.Manager` and no
  longer requires the job queue as an argument. Instead, the manager's `job_queue` will be
  used automatically.
- The `queued_exec` decorator now is re-entrant: decorated functions which are called from
  within the worker thread won't be queued themselves.
- Moved `CustomXepr._wait_stable` to a public method `CustomXepr.waitTemperatureStable`.
- Enforce usage of `exit_customxepr()` to exit.
- Set the default address for email notifications to 'physics-oe-esr-owner@lists.cam.ac.uk'.
- Increased the default timeout for PyVisa communication from 2 sec to 5 sec.

#### Fixed:

- Fixed a bug when plotting aquired results: This was related to the IPython kernel or the
  interactive console using the wrong GUI backend. It now uses the Qt backend.
- Fixed a bug which would cause `XeprData.plot` to fail in case of multiple datasets per
  scan, e.g., for simultanious detection of the in-phase and out-of-phase signals.
- Fixed several Python 3 compatibility issues.

#### Removed:

- Support for Python 2.7. Only Python 3.6 and higher will be supported. Please migrate.

## v2.3.4 (2019-10-09)

Main changes are:

- Possible file names for Xepr data are no longer limited by Xepr's file name restrictions.

config/main.py
# -*- coding: utf-8 -*-
#
# Copyright © Spyder Project Contributors
# Licensed under the terms of the MIT License
# (see spyder/__init__.py for details)

"""
CustomXepr configuration options

Note: Leave this file free of Qt related imports, so that it can be used to
quickly load a user config file
"""
import getpass

# Local import
from config.user import UserConfig

PACKAGE_NAME = 'CustomXepr'
SUBFOLDER = '.%s' % PACKAGE_NAME


# =============================================================================
#  Defaults
# =============================================================================
DEFAULTS = [
            ('Window',
             {
              'x': 0,
              'y': 0,
              'width': 1024,
              'height': 650,
              'splitter': [824, 200],
              }),
            ('CustomXepr',
             {
              'notify_address': [getpass.getuser() + '@cam.ac.uk', ],
              'email_handler_level': 30,
              'temp_wait_time': 120,
              'temperature_tolerance': 0.1
              }),
            (
             'MercuryFeed',
             {
              'MERCURY_ADDRESS': 'TCPIP0::172.20.91.43::7020::SOCKET',
              'temperature_module': 0,
              'gasflow_module': 0,
              'heater_module': 0
              }),
            ('Keithley',
             {
              'KEITHLEY_ADDRESS': 'TCPIP0::192.168.2.121::INSTR',
              'VgStart': 10,
              'VgStop': -60,
              'VgStep': 1,
              'VdList': [-5, -60],
              'VdStart': 0,
              'VdStop': -60,
              'VdStep': 1,
              'VgList': [0, -20, -40, -60],
              'tInt': 0.1,
              'pulsed': False,
              'delay': -1,
              'gate': 'smua',
              'drain': 'smub'
             })
            ]


# =============================================================================
# Config instance
# =============================================================================
# IMPORTANT NOTES:
# 1. If you want to *change* the default value of a current option, you need to
#    do a MINOR update in config version, e.g. from 3.0.0 to 3.1.0
# 2. If you want to *remove* options that are no longer needed in our codebase,
#    or if you want to *rename* options, then you need to do a MAJOR update in
#    version, e.g. from 3.0.0 to 4.0.0
# 3. You don't need to touch this value if you're just adding a new option
CONF_VERSION = '3.0.0'

# Main configuration instance
try:
    CONF = UserConfig('CustomXepr', defaults=DEFAULTS, load=True,
                      version=CONF_VERSION, subfolder=SUBFOLDER, backup=True,
                      raw_mode=True)
except Exception:
    CONF = UserConfig('CustomXepr', defaults=DEFAULTS, load=False,
                      version=CONF_VERSION, subfolder=SUBFOLDER, backup=True,
                      raw_mode=True)
